subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c

```diff
@@ -633,29 +526,326 @@ int nrf_cloud_coap_fetch(const char *resource, const char *query,
+
+	client->cid_saved = false;
+
+	if (nrfc_dtls_cid_is_active(client->sock) && client->authenticated) {
+		err = nrfc_dtls_session_save(client->sock);
+		if (!err) {
+			LOG_DBG("DTLS CID session saved.");
+			client->cid_saved = true;
+		} else {
+			LOG_ERR("Unable to save DTLS CID session: %d", err);
+		}
+	} else {
+		LOG_WRN("Unable to pause CoAP connection.");
+		err = -EACCES;
+	}
+	return err;
+}
+
+int nrf_cloud_coap_transport_resume(struct nrf_cloud_coap_client *const client)
+{
+	if (!client) {
+		return -EINVAL;
+	}
+
+	int err = 0;
+
+	if (client->cid_saved) {
+		err = nrfc_dtls_session_load(client->sock);
+		client->cid_saved = false;
+		if (!err) {
+			LOG_DBG("Loaded DTLS CID session");
+			client->authenticated = true;
+		} else if (err == -EAGAIN) {
+			LOG_INF("Failed to load DTLS CID session");
+		} else if (err == -EINVAL) {
+			LOG_INF("DLTS CID sessions not supported with current modem firmware");
+		} else {
+			LOG_ERR("Error on DTLS CID session load: %d", err);
+		}
+	} else {
+		LOG_WRN("Unable to resume CoAP connection.");
+		err = -EACCES; /* Cannot resume. */
+	}
+
+	return err;
+}
+
+#define PROXY_URI_DL_HTTPS		"https://"
+#define PROXY_URI_DL_HTTPS_LEN		(sizeof(PROXY_URI_DL_HTTPS) - 1)
+#define PROXY_URI_DL_SEP		"/"
+#define PROXY_URI_DL_SEP_LEN		(sizeof(PROXY_URI_DL_SEP) - 1)
+#define PROXY_URI_ADDED_LEN		(PROXY_URI_DL_HTTPS_LEN + PROXY_URI_DL_SEP_LEN)
+
+int nrf_cloud_coap_transport_proxy_dl_opts_get(struct coap_client_option *const opt_accept,
+					       struct coap_client_option *const opt_proxy_uri,
+					       char const *const host, char const *const path)
+{
+	__ASSERT_NO_MSG(opt_accept != NULL);
+	__ASSERT_NO_MSG(opt_proxy_uri != NULL);
+	__ASSERT_NO_MSG(host != NULL);
+	__ASSERT_NO_MSG(path != NULL);
+
+	size_t uri_idx = 0;
+	size_t host_len = strlen(host);
+	size_t path_len = strlen(path);
+
+	opt_accept->code = COAP_OPTION_ACCEPT;
+	opt_accept->len = 1;
+	opt_accept->value[0] = COAP_CONTENT_FORMAT_TEXT_PLAIN;
+
+	opt_proxy_uri->code = COAP_OPTION_PROXY_URI;
+	opt_proxy_uri->len = host_len + path_len + PROXY_URI_ADDED_LEN;
+
+	if (opt_proxy_uri->len > CONFIG_COAP_EXTENDED_OPTIONS_LEN_VALUE) {
+		LOG_ERR("Host and path for CoAP proxy GET is too large: %u bytes",
+			opt_proxy_uri->len);
+		LOG_INF("Increase CONFIG_COAP_EXTENDED_OPTIONS_LEN_VALUE, current value: %d",
+			CONFIG_COAP_EXTENDED_OPTIONS_LEN_VALUE);
+		return -E2BIG;
+	}
+
+	/* We don't want a NULL terminated string, so just copy the data to create the full URI */
+	memcpy(&opt_proxy_uri->value[uri_idx], PROXY_URI_DL_HTTPS, PROXY_URI_DL_HTTPS_LEN);
+	uri_idx += PROXY_URI_DL_HTTPS_LEN;
+
+	memcpy(&opt_proxy_uri->value[uri_idx], host, host_len);
+	uri_idx += host_len;
+
+	memcpy(&opt_proxy_uri->value[uri_idx], PROXY_URI_DL_SEP, PROXY_URI_DL_SEP_LEN);
+	uri_idx += PROXY_URI_DL_SEP_LEN;
+
+	memcpy(&opt_proxy_uri->value[uri_idx], path, path_len);
+
+	LOG_DBG("Proxy URI: %.*s", opt_proxy_uri->len, opt_proxy_uri->value);
+
+	return 0;
+}
```

Comment:
This seems like a whole brand new feature to me? I also don't see it used anywhere

Should this be in a different PR?

Reply:
this is an API that will be used for proxy downloads.
a future PR will use it.